packages/react-dom/src/events/DOMEventResponderSystem.js

```diff
@@ -0,0 +1,197 @@
+  }
+}
+
+DOMEventResponderContext.prototype.dispatchEvent = function(
+  eventName: string,
+  eventListener: EventListener,
+  eventTarget: AnyNativeEvent,
+  discrete: boolean,
+  extraProperties?: Object,
+): void {
+  const eventTargetFiber = getClosestInstanceFromNode(eventTarget);
+  const syntheticEvent = SyntheticEvent.getPooled(
+    null,
+    eventTargetFiber,
+    this.event,
+    eventTarget,
+  );
+  if (extraProperties !== undefined) {
+    copyEventProperties(extraProperties, syntheticEvent);
+  }
+  syntheticEvent.type = eventName;
+  syntheticEvent._dispatchInstances = [eventTargetFiber];
+  syntheticEvent._dispatchListeners = [eventListener];
+
+  let events;
+  if (discrete) {
+    events = this._discreteEvents;
+    if (events === null) {
+      events = this._discreteEvents = [];
+    }
+  } else {
+    events = this._nonDiscreteEvents;
+    if (events === null) {
+      events = this._nonDiscreteEvents = [];
+    }
+  }
+  events.push(syntheticEvent);
+};
+
+DOMEventResponderContext.prototype._runEventsInBatch = function(): void {
+  if (this._discreteEvents !== null) {
+    interactiveUpdates(() => {
+      runEventsInBatch(this._discreteEvents);
+    });
+  }
+  if (this._nonDiscreteEvents !== null) {
+    runEventsInBatch(this._nonDiscreteEvents);
+  }
+};
+
+function createValidEventTypeSet(targetEventTypes): Set<DOMTopLevelEventType> {
+  const eventTypeSet = new Set();
+  // Go through each target event type of the event responder
+  for (let i = 0, length = targetEventTypes.length; i < length; ++i) {
+    const targetEventType = targetEventTypes[i];
+
+    if (typeof targetEventType === 'string') {
+      eventTypeSet.add(((targetEventType: any): DOMTopLevelEventType));
+    } else {
+      if (__DEV__) {
+        warning(
+          typeof targetEventType === 'object' && targetEventType !== null,
+          'Event Responder: invalid entry in targetEventTypes array. ' +
+            'Entry must be string or an object. Instead, got %s.',
+          targetEventType,
+        );
+      }
+      const targetEventConfigObject = ((targetEventType: any): {
+        name: DOMTopLevelEventType,
+        passive?: boolean,
+        capture?: boolean,
+      });
+      eventTypeSet.add(targetEventConfigObject.name);
+    }
+  }
+  return eventTypeSet;
+}
+
+function handleTopLevelType(
+  topLevelType: DOMTopLevelEventType,
+  fiber: Fiber,
+  context: Object,
+): void {
+  const responder: ReactEventResponder = fiber.type.responder;
+  const {props, stateMap} = fiber.stateNode;
+  let validEventTypesForResponder = eventResponderValidEventTypes.get(
+    responder,
+  );
+
+  if (validEventTypesForResponder === undefined) {
+    validEventTypesForResponder = createValidEventTypeSet(
+      responder.targetEventTypes,
+    );
+    eventResponderValidEventTypes.set(responder, validEventTypesForResponder);
+  }
+  if (!validEventTypesForResponder.has(topLevelType)) {
+    return;
+  }
+  let state = stateMap.get(responder);
```

Comment:
The responder can't change because the ReactEventComponent (`fiber.type` here) is effectively frozen as I understand it.

The stateMap lives on the Fiber.

So there is a 1:1 relationship between the map and the responder.

So there can only ever be one value in the stateMap.

Can we just make it a lazily initialized field instead of a Map?